### ISSUE TITLE
Use direct-Helm plugin in the UI

### ISSUE THREAD

**Issue by antgamdia** [kubeapps/kubeapps#3085]
This issue is tracking the progress on making the current UI use the new grpc KubeappsAPIs.
It involves changes to migrate the current UI to use the new API (just displaying the catalog) plus changes in the operations and fields.

## Pending

- [x] [priority/higest] Get the tests to work again. I'd suggest a twofold approach:
    - [x] Ensure every e2e is passing ASAP
    - [x] Start fixing the failing unit tests, if too complex, skip. Prioritize coverage over test quality
- [ ] [priority/high] Implement operations for the Helm actions, namely:
    - [ ] CreateInstalledPackage
    - [ ] UpdateInstalledPackage
    - [ ] DeleteInstalledPackage
- [ ] [priority/medium] Implement the corresponding core operations for each op already implemented in the Helm direct plugin (I'd suggest something quick and then think about the pagination and dups handling later)
- [ ] [priority/low] Rename all the references to "charts" in the code to "package" or sth similar. Do it in a separate PR for reducing noise.

## Done

- [x] Add new operations (tracked at https://github.com/kubeapps/kubeapps/issues/3146)
- [x] Add minor changes "on the fly" (tracked at https://github.com/kubeapps/kubeapps/issues/3165)
  - Missing fields tracked at https://github.com/kubeapps/kubeapps/issues/3165#issuecomment-882943900
  - Missing secondary operations tracked at https://github.com/kubeapps/kubeapps/issues/3165#issuecomment-882944035
- [x] Perform the UI migration 
  - Required extra changes (eg, testing, refactor, etc) tracked at https://github.com/kubeapps/kubeapps/issues/3165#issuecomment-882945858
  - Actual UI migration divided per use case tracked at https://github.com/kubeapps/kubeapps/issues/3165#issuecomment-882950182

- [x]  Check, quickly, the current UI requirements (in terms of field required in the data model) using the non-grpc calls.
  - https://github.com/kubeapps/kubeapps/issues/3085#issuecomment-876196248
- [x] Investigate and add support for performing grpc calls from the client
  - https://github.com/kubeapps/kubeapps/pull/3155 + discussion in this issue

_(updated july 20th)_

**Comment by antgamdia** [comment]
(Reposting it here for completeness from https://github.com/kubeapps/kubeapps/issues/2993#issuecomment-872414596)

I'm now addressing the UI changes. Since it is a complex endeavor, let's break this task down into three different milestones:

1.  Proof of concept just changing the endpoints and data models used in the client (no feature flag yet). The outcome of this phase should be a list of requirements for the API responses. For instance: a missing field o some logic that should be addressed as part of the backed. It will allow other people to continue to work in parallel.
2. Evaluate a major change in the API client to directly use gRPC w/ Typescript. It involves:
    - Check if the current proxying approach works for gRPC or if need to add a new http2 Nginx listener to proxy pass it.
        - Update: nginx cannot multiplex HTTP 1.1 and 2 without custom logic. I've tried adding a new location and using `grpc_pass`, but no success. I've ended up creating a new `server { listen <another port> http2; ...` (plus a new k8s  svc port mapping) and it works. I'll go this way until we come up with any other idea.
    - Implement a proof of concept (take into account the pagination)
3. Add a feature flag (or sth similar) for landing code quickly.

### Raw issues while performing UI changes

_Placeholder section for tracking (raw) the issues I'm finding; later on, I'll create a proper summary_

- [ ] Add `repo` to `GetAvailablePackageSummaries` and `GetAvailablePackageDetails`.
    - **Status**: to be considered as an optional field on the summary and detail. 
    - **Core API relationship**: this field is not available for every plugin
    - **ETA**: until we define the package repository messages (including a package repository reference)
    - **Reason**: it is required for filtering by repo in the UI. This functionality is used in both the catalog view (as a filter) and the AppRepo view (as a link to the catalog).

- [ ] Add `latestAppVersion` to `GetAvailablePackageSummaries`.
    - **Status**: TBD
    - **Core API relationship**: TBD
    - **ETA**: TBD
    - **Reason**:  appears in the code, but checking why it is needed still pending.

- [ ] Add `category` to `GetAvailablePackageSummaries`.
    - **Status**: TBD
    - **Core API relationship**: TBD
    - **ETA**: TBD
    - **Reason**: it is required for filtering by category in the UI. Is it really needed? It is sth that almost only Bitnami provides, no? Check if Carvel also offers (or will offer) categories. 

- [ ] Issue when querying a namespace in the API, 401: `{"code":16,"message":"The current user has no access to the namespace \"default\""}` Check tomorrow if it is really a problem with the can-I endpoint or is just a personal RBAC issue (though... I'm cluster.admin)

#### Current status
![image](https://user-images.githubusercontent.com/11535726/124171788-4b255c80-daa9-11eb-9960-75941f8a47f0.png)


#### Thoughts for the Helm-less version

- No category and repo filters whatsoever ? 
- Top filter by "plugin" that triggers package-specific filters like category/repo ?
    - Like this: 
![image](https://user-images.githubusercontent.com/11535726/124173401-4661a800-daab-11eb-9f83-63ab900025e2.png)

- ID/name collision between plugins, are we handling this no? Consider names like "aaa/bbb"

**Comment by antgamdia** [comment]
Well, some updates here (finally!): 

### The problem 

Before starting any UI effort, I've been investigating how we are going to consume the gRPC API. Unfortunately, our options are pretty reduced considering this premise: "is not straightforward to expose the gRPC API through our nginx proxy and have clients (any web browser) able to consume it" 

There are some code generators `.proto`->`.ts`, which is great. They generate both types/interfaces and client code, but this code is not really useful for us.
**Example**: https://github.com/stephenh/ts-proto  (it generates `Twirp` (doesn't support Go), `grpc-web` (will cover it later), `grpc-js` (just nodejs backends) and `nestjs` (just nodejs backends)

My next step was trying to answer this question: can I just consume a gRPC endpoint from my browser with a simple library. Unfortunately again, the answer is no. Here is more information [1].

At this point, `grpc-web` arises again. What is it? As the website states: _due to browser limitation, the Web client library implements a different protocol than the native gRPC protocol. This protocol is designed to make it easy for a proxy to translate between the protocols as this is the most likely deployment model._ [2]

(So, my initial premise got confirmed now... it wasn't that easy to just serve the grpc via nginx.)

`grpc-web` is an alternative to the `grpc-gateway` we are currently using [1]: 

![image](https://user-images.githubusercontent.com/11535726/125368745-40a57580-e37b-11eb-9f9d-8a51de893cf7.png)
_Image credits go to improbable.io_ 

So, what if we may want to use `grpc-web`? Well, in that case, we would need to face an additional problem: configuring nginx properly. The default out-of-the-box approach is using Envoy instead of nginx; some people point out certain solutions, though [3]. 

This overall rationale is also commented on here [4].

### Conclusions 

If we happen to want to go with a fully grpc solution end to end, perhaps we should re-think the backend side again and think of migrating from `grpc-gateway` to `grpc-web` with some hacky configuration (although I'm really willing to give Envoy a try; it is used in plenty of API GW solutions).
Otherwise, we can assume a typical REST-driven frontend (as we already had) and just using the endpoints already offered by `grpc-gateway`.
Note also that there are some benchmarks [5,6] that show (albeit not statistically representative) the `grpc-gateway` might be a 30% slower compared to the `grpc-web`.

My thoughts here: given that we have plenty of work from the UI revamping side, adding more complexity here perhaps is not worthy. I'd rather go with the `grpc-gateway` for now, which only would imply a change in the endpoints we currently are calling.

See what you think, happy to consider more viewpoints here.

[1] https://www.improbable.io/blog/grpc-web-moving-past-restjson-towards-type-safe-web-apis
[2] https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-WEB.md
[3] https://stackoverflow.com/questions/54440325/can-we-use-nginx-proxy-for-grpc-web-integration-without-docker-dependency
[4] https://code.cash.app/protoc-gen-grpc-gateway-ts-clean-idiomatic-typescript-for-grpc-gateway
[5] https://github.com/thinhdanggroup/benchmark-grpc-web-gateway
[6] https://medium.com/@thinhda/compare-grpc-web-with-grpc-gateway-fa1e2acdf29f

**Comment by absoludity** [comment]
> Well, some updates here (finally!):

Great, thanks for the investigation! It sounds like you've got a good grasp of the different options, and it's very helpful to read through to get back up-to-speed. I'll sprinkle some comments inline, I think there's one option you may have missed that you might want to compare with the one from your conclusion.

> ### The problem
> 
> Before starting any UI effort, I've been investigating how we are going to consume the gRPC API. Unfortunately, our options are pretty reduced considering this premise: "is not straightforward to expose the gRPC API through our nginx proxy ...

This part would be straight-forward I think? In that, we could (if we needed to) proxy any `application/grpc` request that hit our frontend through to the kubeapps-apis service, just like any other content-type based proxy rule? But that won't help because....

> ...and have clients (any web browser) able to consume it"

Right, browsers don't speak gRPC natively. Any gRPC client would be able to consume the grpc service sitting behind nginx, but a web browser isn't one of those.

> 
> There are some code generators `.proto`->`.ts`, which is great. They generate both types/interfaces and client code, but this code is not really useful for us.

I think the `protoc-gen-grpc-gateway-ts` that you found looks very useful? Or do you just mean the specific ones you mention below aren't so useful for us?

> **Example**: https://github.com/stephenh/ts-proto (it generates `Twirp` (doesn't support Go), `grpc-web` (will cover it later), `grpc-js` (just nodejs backends) and `nestjs` (just nodejs backends)

Right, `grpc-web` is the main one, as you've found, but `grpc-web` requests need to be translated before it gets to the grpc server. And there are a number of ways that can be done (more below).

> (So, my initial premise got confirmed now... it wasn't that easy to just serve the grpc via nginx.)

Well, it's easy to serve the grpc via nginx, but that doesn't help a web client to consume it (I assume that's what you mean?). That's one of the reasons that we originally added grpc-gateway so that the API would also be exposed via http as a baseline (similar to other projects within VMware). But it would be really nice if we are able to use a generated client and types from the protobuf files grpc from the browser as well... let's look at the options you've found and one other.

> So, what if we may want to use `grpc-web`? Well, in that case, we would need to face an additional problem: configuring nginx properly.

Again, it's not clear to me why you think there's some issue with nginx proxying grpc or grpc-web requests. Proxying them should be no issue for any proxy (based on content type), so I can only guess that you mean not just proxying them, but also doing the translation from grpc-web to grpc? In which case, yes, we wouldn't want nginx to be doing that translation. Google's implementation of gRPC web used to rely on an nginx-extension to do the translation, but no longer does (they switched to an envoy filter).

A really good summary of the current state, which you may have read (I just don't see it in your list of references) is https://grpc.io/blog/state-of-grpc-web/ (though even it is 2 years old now).

> The default out-of-the-box approach is using Envoy instead of nginx; some people point out certain solutions, though [3].

If you're using a separate proxy service for the translation of `grpc-web`->`grpc`, then yes, Envoy is what you'll need to use (both the improbable and google implementations of grpc-web now have envoy filters, and finally, they both work with each other). But that's not the only option, afaict? Improbable also provide a go library `grpcweb` that you can import in your existing go grpc server and just wraps your grpc service to handle translating any `grpc-web` requests for you before passing them to your server:

https://github.com/improbable-eng/grpc-web/tree/master/go/grpcweb

With that, there's no envoy or separate translation-proxy required.

> 
> This overall rationale is also commented on here [4].

Note that this isn't an overall rationale for grpc-web, but rather the rational for why cash.app decided *not* to use grpc-web and instead created their protoc-gen-grpc-gateway-ts , I think? (which is different, since I assume it's a typed client that is using the grpc-gateway http api, which is also quite neat). I was just confused with this comment in the context where you seemed to be discussing grpc-web.

> ### Conclusions
> 
> If we happen to want to go with a fully grpc solution end to end, perhaps we should re-think the backend side again and think of migrating from `grpc-gateway` to `grpc-web` with some hacky configuration (although I'm really willing to give Envoy a try; it is used in plenty of API GW solutions).

So if we do want to go with a fully grpc solution end to end, I don't think we do need to use envoy or add any extra proxy? All we'd need to do is update the nginx config to proxy (not translate) grpc requests, which is simple enough, and wrap our go grpc server with the improbable `grpcweb` wrapper. We can still leave grpc-gateway there so we have an http API for those that may want to use that (yes, we still need to define/update proper URL endpoints). We'd then have a single server port which is muxing grpc-web, grpc and http requests.

> Otherwise, we can assume a typical REST-driven frontend (as we already had) and just using the endpoints already offered by `grpc-gateway`.

I assume you mean with the generated client and types from `protoc-gen-grpc-gateway-ts` rather than just interacting with the api directly, but yep, we could.

> Note also that there are some benchmarks [5,6] that show (albeit not statistically representative) the `grpc-gateway` might be a 30% slower compared to the `grpc-web`.

Yep, and we would benefit from that if use the improbable `grpc-web` wrapper together with the improbable typescript client. We'd be including grpc-gateway only so we have an http version available (for free) so other potential clients of the kubeapps apis server have that option. 

> 
> My thoughts here: given that we have plenty of work from the UI revamping side, adding more complexity here perhaps is not worthy. I'd rather go, for now, with the `grpc-gateway` + generating ts code with `protoc-gen-grpc-gateway-ts` [4], which only would imply a change in the endpoints we currently are calling.
> 
> See what you think, happy to consider more viewpoints here.

So I'm guessing that conclusion was drawn without seeing the `grpcweb` wrapper, which doesn't add more complexity (ie. we don't need to add yet another separate proxying service to do the translation). Take a look and see what you think. I'm OK with either solution, depending what you find. Both solutions don't require any extra services/complexity and would give us a typescript client with types for the messages to use in the dashboard, which is the main thing I'm keen for (ie. I don't want us writing our own interactions with the http api server). One is sending the messages via grpc-web directly (ie. no translation to http requests), the other is sending the messages to http endpoints for gateway to translate, but there's translation involved both ways.

Keen to hear your final thoughts after reading the above, but I'll be happy with either of those. I hope it's been a fun and useful bit of research for you!

> 
> [1] https://www.improbable.io/blog/grpc-web-moving-past-restjson-towards-type-safe-web-apis
> [2] https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-WEB.md
> [3] https://stackoverflow.com/questions/54440325/can-we-use-nginx-proxy-for-grpc-web-integration-without-docker-dependency
> [4] https://code.cash.app/protoc-gen-grpc-gateway-ts-clean-idiomatic-typescript-for-grpc-gateway
> [5] https://github.com/thinhdanggroup/benchmark-grpc-web-gateway
> [6] https://medium.com/@thinhda/compare-grpc-web-with-grpc-gateway-fa1e2acdf29f

**Comment by antgamdia** [comment]
> This part would be straight-forward I think? In that, we could (if we needed to) proxy any `application/grpc` request that hit our frontend through to the kubeapps-apis service, just like any other content-type based proxy rule? But that won't help because....

Yes, perhaps I didn't get to explain very well. I mean that is not trivial serving it in the same way we currently do (I mean, just adding the grpc, but with the whole REST via http1 thing also working.
That's why I just exposed another separate port (just for grpc) in my local setup: https://github.com/antgamdia/kubeapps/blob/WIP_grpc_api_frontend/chart/kubeapps/templates/frontend/configmap.yaml#L189

> I think the `protoc-gen-grpc-gateway-ts` that you found looks very useful? Or do you just mean the specific ones you mention below aren't so useful for us? 

Yes, it seems to be so. I didn't manage yesterday to also look at it, but it's something I want to address today.

> Well, it's easy to serve the grpc via nginx, but that doesn't help a web client to consume it (I assume that's what you mean?). 

Yes, that's what I meant.

> You mean not just proxying them, but also doing the translation from grpc-web to grpc? In which case, yes, we wouldn't want nginx to be doing that translation

Yep, exactly, that is the issue.

> If you're using a separate proxy service for the translation of `grpc-web`->`grpc`, then yes, Envoy is what you'll need to use (both the improbable and google implementations of grpc-web now have envoy filters, and finally, they both work with each other). But that's not the only option, afaict? Improbable also provide a go library `grpcweb` that you can import in your existing go grpc server and just wraps your grpc service to handle translating any `grpc-web` requests for you before passing them to your server:
> https://github.com/improbable-eng/grpc-web/tree/master/go/grpcweb

Yes, it is an additional option. I haven't look at it closely because I don't know if `grpcweb` gets along with `grpc-gateway`. Certainly, it is a good option we really want e2e grpc backend. 


> Note that this isn't an overall rationale for grpc-web, but rather the rational for why cash.app decided _not_ to use grpc-web and instead created their protoc-gen-grpc-gateway-ts , I think? (which is different, since I assume it's a typed client that is using the grpc-gateway http api, which is also quite neat). I was just confused with this comment in the context where you seemed to be discussing grpc-web.

Yes, it is indeed their rationale on their decision (which could also be useful for us).

> So if we do want to go with a fully grpc solution end to end, I don't think we do need to use envoy or add any extra proxy? All we'd need to do is update the nginx config to proxy (not translate) grpc requests, which is simple enough, and wrap our go grpc server with the improbable `grpcweb` wrapper. We can still leave grpc-gateway there so we have an http API for those that may want to use that (yes, we still need to define/update proper URL endpoints). We'd then have a single server port which is muxing grpc-web, grpc and http requests.


Mmm, fair enough! Poor mux server... we are overloading its duties in each PR haha.  It is certainly a good option to be considered.

> I assume you mean with the generated client and types from `protoc-gen-grpc-gateway-ts` rather than just interacting with the api directly, but yep, we could.

Yes, that's what I meant; using the rest API but leveraging the autogenerated client and types.

> Yep, and we would benefit from that if use the improbable `grpc-web` wrapper together with the improbable typescript client. We'd be including grpc-gateway only so we have an http version available (for free) so other potential clients of the kubeapps apis server have that option.

> So I'm guessing that conclusion was drawn without seeing the `grpcweb` wrapper, which doesn't add more complexity 

More or less; whereas I hadn't explicitly considered, I'm not 100% sure I'm choosing another solution, let me explain: given that both solutions involve an ulterior transformation, I don't know the real benefit from going all-in with the grpc way right now. I haven't checked if the performance of `gprweb` is better than the plain `grpc-gateway` (might be, though), so I can't really tell.

However, I do see some problems (now) if we go this way (grpcweb): as we want to support both the legacy and pluggable backend with the same UI (plus feature flags or whatever), adding a completely different API client perhaps is gonna add more complexity here (note that we already have a well-tested `axiosWithAuth` client, for instance).

I want to personally give `gprcweb` a try and see how is it so I can make an informed decision, however, I don't see currently it worths implementing another client in the frontend code (as we have to (still) support the legacy/current approach).
Off the top of my head, what I would suggest is:

- **Backend**: adding the `grpcweb` wrapper, which will enable other frontends/clients to use the grpc-web protocol instead of the normal (aka "rest") one.
- **Fronted**: continue using the rest client via `grpc-gateway` with autogenerated types and clients with `protoc-gen-grpc-gateway-ts`. This way, changing from one service to another is just a matter of changing the endpoint+type (not the client itself).

> Keen to hear your final thoughts after reading the above, but I'll be happy with either of those. I hope it's been a fun and useful bit of research for you!

Let me first play around with `grpcweb` (plus generating code for grpc-web) and how much complexity it involves. Keep you posted!

**Comment by antgamdia** [comment]
> Let me first play around with `grpcweb` (plus generating code for grpc-web) and how much complexity it involves. Keep you posted!

Well...  consuming the API seems to be much simpler than expected (using the `ts-proto` generator instead, [see a full example here](https://github.com/stephenh/ts-proto/blob/main/integration/grpc-web/client-ts.ts)).

```js
    // Create the client
    const rpc = new GrpcWebImpl("http://localhost:50052", {
      transport: NodeHttpTransport(),
      debug: false,
      metadata: new grpc.Metadata({ SomeHeader: "bar" }),
    });
    const client = new HelmPackagesServiceClientImpl(rpc);
    
    // Use it
    return await client.GetAvailablePackageSummaries({ context: { namespace: "kubeapps" } });
```

I still have to deal with the authentication, but I think it's just a matter of checking the examples and the generator options.

Regarding the latencies, they are pretty similar, I'm not able to find any decisive difference. However, I must reckon my local dev setup is not really suitable for giving a confident answer with regard to the performance. 

- Using grpc-web
  - Endpoint: `http://localhost:50052/kubeappsapis.plugins.helm.packages.v1alpha1.HelmPackagesService/GetAvailablePackageSummaries`
  - Time elapsed: ~80-90ms

- Using grpc-gateway
  - Endpoint: `http://localhost:50051/core/packages/v1alpha1/availablepackagesummaries?context.namespace=kubeapps` 
  - Time elapsed: ~80-90rms


For adding the grpcweb, all I had to do was adding a new listener in another port also serving this protocol. Then, we would have to add a `grpc_pass` directive in nginx.

(btw, I saw [this similar approach here](https://github.com/FleekHQ/space-daemon/blob/master/grpc/grpc.go), it's a neat solution for keeping everything clean and tidy. As the number of listeners increases, perhaps we could add another task for refactoring it a little bit).

So, at this very moment (having implemented the PoC; if curious, [it's in one of my fork's branches](https://github.com/antgamdia/kubeapps/blob/WIP_grpc_api_frontend/cmd/kubeapps-apis/server/server.go)), I don't mind going either way.
Perhaps the quasi-end-to-end-grpc way is more consistent with our movement towards grpc. Besides, nothing stops other clients from using the REST endpoints if they want.

From the UI code's point of view, given that we have to support both approaches right now, is not too bad to have two separate clients: the grpc and the HTTP one.

**Backend**: add the grpcweb wrapper, which will enable other frontends/clients to use the grpc-web protocol instead of the normal (aka "rest") one, plus adapt the chart to serve an additional grpcweb port.

**Fronted**: start using grpc calls also here leveraging the autogenerated code by `ts-proto`. This solution requires: 1) the autogenerated types; 2) the autogenerated logic to en/decode grpc calls; 3) a grpcweb client + a grpcweb transport (ie, just a couple of extra deps)

**Pending things**: investigate how can we perform authenticated calls.

**Personal concerns**: afaik, exposing grpc via http2 requires https, I don't know if would have any problems when using http as the transport layer. Also, I don't know how this whole setup is gonna behave when using an ingress on the top of it.

So, my conclusion so far: if, after exposing everything in the chart and using a non-dev environment*, it works. I'd go ahead with the grpc approach. Otherwise, I'll take the conservative way, and just use the grpc-gateway+generated types.

* right now I'm just hardcoding my localhost URL after running the kubeappsapi binary by hand and proxying UI requests, so definitely, not a prod-alike env.

Keep you posted! I hope in my next post I could have a final decision (sorry for the delay, I just want to really make sure that the chosen option is the best for our use case).

**Comment by absoludity** [comment]
> Well... consuming the API seems to be much simpler than expected (using the `ts-proto` generator instead, [see a full example here](https://github.com/stephenh/ts-proto/blob/main/integration/grpc-web/client-ts.ts)).
> 
> ```js
>     // Create the client
>     const rpc = new GrpcWebImpl("http://localhost:50052", {
>       transport: NodeHttpTransport(),

Isn't that transport specifically if you're using grpcweb from nodejs? Ah, perhaps you were just testing from nodejs (as by default in a web browser context it should be using `grpc.CrossBrowserHttpTransport` I think, as per [the docs](https://github.com/improbable-eng/grpc-web/blob/master/client/grpc-web/docs/transport.md#crossbrowserhttptransport))

>       debug: false,
>       metadata: new grpc.Metadata({ SomeHeader: "bar" }),
>     });
>     const client = new HelmPackagesServiceClientImpl(rpc);
>     
>     // Use it
>     return await client.GetAvailablePackageSummaries({ context: { namespace: "kubeapps" } });
> ```
> 
> I still have to deal with the authentication, but I think it's just a matter of checking the examples and the generator options.

Actually, auth could be a bit more interesting, in that JS in the browser doesn't have access to the authentication (the oauth2-proxy cookie is an http-only cookie), so it's not something we'll be setting in the request. Similar to any other request from the dashboard, oauth2-proxy will convert the http-only cookie into ot `Authorization` header on the way through to the kubeapps apis server. Hopefully grpc automatically converts that to the corresponding metadata, but I'm doubtful (since you can set the metadata in the client). And if not, we'll need to extract the HTTP header and insert it into the metadata ourselves before handling the grpc request.  

> 
> Regarding the latencies, they are pretty similar, I'm not able to find any decisive difference. However, I must reckon my local dev setup is not really suitable for giving a confident answer with regard to the performance.

> For adding the grpcweb, all I had to do was adding a new listener in another port also serving this protocol.

I'm not even sure why we'd need another listener, given that we can wrap the existing `grpcLis` listener (it already handles all `application/grpc` content and the grpcweb wrapper is designed exactly to filter the `grpc-web` traffic from the `grpc` ones).

> Then, we would have to add a `grpc_pass` directive in nginx.

The `grpc_pass` directive would need to be passing both `application/grpc` and `application/grpc-web` anyway, so it may as well do it to the one port. Unless there's a reason not to?

> 
> (btw, I saw [this similar approach here](https://github.com/FleekHQ/space-daemon/blob/master/grpc/grpc.go), it's a neat solution for keeping everything clean and tidy. As the number of listeners increases, perhaps we could add another task for refactoring it a little bit).

Yeah. I'd hoped we could just add a

```
grpcwebLis := mux.MatchWithWriters(cmux.HTTP2MatchHeaderFieldSendSettings("content-type", "application/grpc-web"))
```

and used that with the wrapped grpcweb server, but maybe that's not possible (grpcweb's wrapped server appears to expose the http.Handler interface only). Not sure.

> From the UI code's point of view, given that we have to support both approaches right now, is not too bad to have two separate clients: the grpc and the HTTP one.

Yes, I think that's fine for the new and the old.

> 
> **Personal concerns**: afaik, exposing grpc via http2 requires https, I don't know if would have any problems when using http as the transport layer. Also, I don't know how this whole setup is gonna behave when using an ingress on the top of it.

Oh? Interested to see more about that. I didn't think I'd had issues with grpc over http in the past. I don't see why ingress would change anything either, but keen to hear when you dig more.

> Keep you posted! I hope in my next post I could have a final decision (sorry for the delay, I just want to really make sure that the chosen option is the best for our use case).

It's not a delay, it's good investigation that will save us time later :) Great stuff.

**Comment by antgamdia** [comment]
> Isn't that transport specifically if you're using grpcweb from nodejs? Ah, perhaps you were just testing from nodejs (as by default in a web browser context it should be using grpc.CrossBrowserHttpTransport I think, as per the docs)

No, I just copy-n-pasted an example out there. I've read the docs (thanks for the pointer!) and now it's changed.

> Hopefully grpc automatically converts that to the corresponding metadata, but I'm doubtful (since you can set the metadata in the client). And if not, we'll need to extract the HTTP header and insert it into the metadata ourselves before handling the grpc request.

(Finally) good news here. The proper headers are being automatically passed through the chain where it finally reaches our code where we extract the token. I have tested this behavior in both token authentication and oidc (w/ pinniped)  auth.

> I'm not even sure why we'd need another listener (...) Yeah. I'd hoped we could just add a (...) and used that with the wrapped grpcweb server, but maybe that's not possible (grpcweb's wrapped server appears to expose the http.Handler interface only). Not sure.

Yeah! It works!!  I just added it as a handler of the existing `httpSrv`  as an `http.HandlerFunc`

> Oh? Interested to see more about that. I didn't think I'd had issues with grpc over http in the past. I don't see why ingress would change anything either, but keen to hear when you dig more.

And now the not-so-good news. When I said "exposing grpc via http2 requires https", I really meant the following:
The HTTP/2 protocol when not using TLS is called `h2c` . Well, [nginx does -not- support both HTTP/1.1 AND h2c in the same socket](https://trac.nginx.org/nginx/ticket/816). Given that we do not configure any TLS layer in this nginx, as a result, we cannot expose our grpc via http2 (h2c) in the same port as it would require https.

Consequences? Not really. We certainly can perform grpc calls via grpcweb in a single port (`/apis`, as usual). However, these calls use HTTP/1.1 instead of HTTP/2... not a big deal perhaps, I don't have any benchmark numbers handy.

Idea: it is not possible to multiplex our calls as I said before, however, nothing stops us from adding a new port that listens grpc via http2 for those clients (like the `grpcurl` tool). It would be something configurable and conditional, like `Values.kubeappsapis.exposeGrpcHttp2: true`.
That said, it is not required for our goal, if you think it's useful, we could just create an issue for now,

Having played around with the configurations and environments, I've sent a PR with my current work (https://github.com/kubeapps/kubeapps/pull/3155). If it happens to make sense, we could start performing the actual changes in the frontend.

**Comment by absoludity** [comment]
Great work as always!

> And now the not-so-good news. When I said "exposing grpc via http2 requires https", I really meant the following:
> The HTTP/2 protocol when not using TLS is called `h2c` . Well, [nginx does -not- support both HTTP/1.1 AND h2c in the same socket](https://trac.nginx.org/nginx/ticket/816). Given that we do not configure any TLS layer in this nginx, as a result, we cannot expose our grpc via http2 (h2c) in the same port as it would require https.
> 
> Consequences? Not really. We certainly can perform grpc calls via grpcweb in a single port (`/apis`, as usual). However, these calls use HTTP/1.1 instead of HTTP/2... not a big deal perhaps, I don't have any benchmark numbers handy.

Right, so on a call I think we agreed that:

- Given that currently there is no change for the user/client (the client will just use 1.1 without issue), we're OK with this for now, and
- When we do want to ensure http/2 is used, we should upgrade our config to ensure a TLS connection between nginx and kubeapps apis (which we want to do anyway for all internal services eventually).

Let me know if I missed something.

**Comment by antgamdia** [comment]
Closing in favor of https://github.com/kubeapps/kubeapps/issues/3340
